src/Units/Unit.java

```diff
@@ -91,37 +92,30 @@ public double getBaseDamage(UnitModel target, int range, boolean moved)
     // if we have no weapons, we can't hurt things
     if( weapons == null )
       return 0;
+    double maxDamage = 0;
     for( int i = 0; i < weapons.length; i++ )
     {
       // If the weapon isn't mobile, we shouldn't be able to fire if we moved.
       if( moved && !weapons[i].model.canFireAfterMoving )
       {
         continue;
       }
-      double damage = weapons[i].getDamage(target, range);
-      if( damage != 0 )
+      double currentDamage = weapons[i].getDamage(target, range);
+      if( maxDamage < currentDamage )
       {
-        return damage;
+        maxDamage = currentDamage;
+        currentWeapon = weapons[i];
       }
     }
-    return 0;
+    return maxDamage;
   }
 
   // for the purpose of letting the unit know it has attacked.
   public void fire(final Unit defender)
   {
-    UnitEnum target = defender.model.type;
-    int i = 0;
-    for( ; i < weapons.length; i++ )
-    {
-      if( weapons[i].getDamage(target) != 0 )
-      {
-        break;
-      }
-    }
-    if( i == weapons.length )
+    if( null == currentWeapon )
       System.out.println("In " + model.name + "'s fire(): no valid weapon found");
-    weapons[i].fire();
+    currentWeapon.fire();
```

Comment:
Ehhh, like you pointed out, the disconnect between `getBaseDamage` and `fire` makes me uneasy. Probably worth looking into an alternate approach here.

Reply:
My problem is that I'm not sure what to do about it.
fire() doesn't have the proper information to make that determination, and it would be re-doing getDamage()'s work anyway.
I suppose we could add another boolean to getDamage that tells the unit "hey we're firing for real this time"?

Reply:
Made a new pull request to try and address this. Lemme know what you think.